protocol/app/ante/market_update_test.go

```diff
@@ -0,0 +1,502 @@
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+
+			if got != tt.want {
+				t.Errorf("IsMarketUpdateTx() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestValidateMarketUpdateDecorator_AnteHandle(t *testing.T) {
+	type marketPerpPair struct {
+		market prices_types.MarketParam
+		perp   perpetualtypes.Perpetual
+	}
+
+	type args struct {
+		msgs        []sdk.Msg
+		simulate    bool
+		marketPerps []marketPerpPair
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "do nothing for non-market map messages - simulate",
+			args: args{
+				msgs: []sdk.Msg{
+					&banktypes.MsgSend{
+						FromAddress: constants.BobAccAddress.String(),
+						ToAddress:   constants.AliceAccAddress.String(),
+						Amount: []sdk.Coin{
+							sdk.NewCoin(assets.AssetUsdc.Denom, sdkmath.NewInt(1)),
+						},
+					},
+					&banktypes.MsgSend{
+						FromAddress: constants.BobAccAddress.String(),
+						ToAddress:   constants.AliceAccAddress.String(),
+						Amount: []sdk.Coin{
+							sdk.NewCoin(assets.AssetUsdc.Denom, sdkmath.NewInt(1)),
+						},
+					},
+				},
+				simulate: true,
+			},
+			wantErr: false,
+		},
+		{
+			name: "do nothing for non-market map messages",
+			args: args{
+				msgs: []sdk.Msg{
+					&banktypes.MsgSend{
+						FromAddress: constants.BobAccAddress.String(),
+						ToAddress:   constants.AliceAccAddress.String(),
+						Amount: []sdk.Coin{
+							sdk.NewCoin(assets.AssetUsdc.Denom, sdkmath.NewInt(1)),
+						},
+					},
+					&banktypes.MsgSend{
+						FromAddress: constants.BobAccAddress.String(),
+						ToAddress:   constants.AliceAccAddress.String(),
+						Amount: []sdk.Coin{
+							sdk.NewCoin(assets.AssetUsdc.Denom, sdkmath.NewInt(1)),
+						},
+					},
+				},
+				simulate: false,
+			},
+			wantErr: false,
+		},
+		{
+			name: "reject for multiple market messages - simulate",
+			args: args{
+				msgs: []sdk.Msg{
+					&mmtypes.MsgUpsertMarkets{
+						Authority: constants.BobAccAddress.String(),
+						Markets:   nil,
+					},
+					&mmtypes.MsgUpsertMarkets{
+						Authority: constants.BobAccAddress.String(),
+						Markets:   nil,
+					},
+				},
+				simulate: true,
+			},
+			wantErr: true,
+		},
+		{
+			name: "reject for multiple market messages",
+			args: args{
+				msgs: []sdk.Msg{
+					&mmtypes.MsgUpsertMarkets{
+						Authority: constants.BobAccAddress.String(),
+						Markets:   nil,
+					},
+					&mmtypes.MsgUpsertMarkets{
+						Authority: constants.BobAccAddress.String(),
+						Markets:   nil,
+					},
+				},
+				simulate: false,
+			},
+			wantErr: true,
+		},
+		{
+			name: "accept a single message with no markets",
+			args: args{
+				msgs: []sdk.Msg{
+					&mmtypes.MsgUpsertMarkets{
+						Authority: constants.BobAccAddress.String(),
+						Markets:   nil,
+					},
+				},
+				simulate: false,
+			},
+			wantErr: false,
+		},
+		{
+			name: "accept a single message with no cross markets",
```

Comment:
nit: Add separate test where the market exists, in the perp pairs, but isn't cross (is isolated). This test case only covers a market that doesn't exist in the set of perps.